pkg/imgpkg/bundle/bundle_images_lock.go

```diff
@@ -9,53 +9,83 @@ import (
 	"io"
 	"io/ioutil"
 	"path/filepath"
+	"strings"
 	"sync"
 
+	regname "github.com/google/go-containerregistry/pkg/name"
 	regv1 "github.com/google/go-containerregistry/pkg/v1"
 	"github.com/google/go-containerregistry/pkg/v1/types"
 	"github.com/k14s/imgpkg/pkg/imgpkg/lockconfig"
 	"github.com/k14s/imgpkg/pkg/imgpkg/util"
 )
 
-func (o *Bundle) AllImagesLock(concurrency int) (*ImagesLock, error) {
+func (o *Bundle) AllImagesRefs(concurrency int, logger util.LoggerWithLevels) ([]*Bundle, ImagesRef, error) {
 	throttleReq := util.NewThrottle(concurrency)
-	return o.buildAllImagesLock(&throttleReq, &processedImages{processedImgs: map[string]struct{}{}})
+	bundles, imagesLock, err := o.buildAllImagesLock(&throttleReq, &processedImages{processedImgs: map[string]struct{}{}}, logger)
+	if err != nil {
+		return nil, ImagesRef{}, err
+	}
+
+	imagesRef := imagesLock.ImagesRefs()
+	// Ensure that the correct IsBundle flag is provided.
+	// This loop needs to happen because we skipped some images for some bundle, and only at this point we have
+	// the full list of ImagesRef created and can fill the gaps inside each bundle
+	for _, bundle := range bundles {
+		for _, ref := range bundle.ImagesRef() {
+			imgRef, found := imagesRef.ImageRef(ref.Image)
+			if !found {
+				panic(fmt.Sprintf("Internal inconsistency: The Image '%s' cannot be found in the total list of images", ref.Image))
+			}
+
+			bundle.AddImagesRef(imgRef)
+		}
+	}
+
+	return bundles, imagesRef, err
 }
 
-func (o *Bundle) buildAllImagesLock(throttleReq *util.Throttle, processedImgs *processedImages) (*ImagesLock, error) {
+func (o *Bundle) buildAllImagesLock(throttleReq *util.Throttle, processedImgs *processedImages, logger util.LoggerWithLevels) ([]*Bundle, *ImagesLock, error) {
 	img, err := o.checkedImage()
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
-	imagesLock, err := o.imagesLockReader.Read(img)
+	resultImagesLock := NewImagesLock(lockconfig.ImagesLock{}, o.imgRetriever, o.Repo())
+
+	imageRefs, err := o.fetchImagesRef(img, resultImagesLock, logger)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
-	resultImagesLock := NewImagesLock(lockconfig.ImagesLock{}, o.imgRetriever, o.Repo())
-	currentImagesLock := NewImagesLock(imagesLock, o.imgRetriever, o.Repo())
+	bundles := []*Bundle{o}
 
-	errChan := make(chan error, len(imagesLock.Images))
+	errChan := make(chan error, len(imageRefs))
 	mutex := &sync.Mutex{}
 
-	for _, image := range currentImagesLock.ImageRefs() {
+	for _, image := range imageRefs {
+		o.AddImagesRef(image)
+
 		if skip := processedImgs.CheckAndAddImage(image.Image); skip {
 			errChan <- nil
 			continue
 		}
 
 		image := image.DeepCopy()
 		go func() {
-			imgsLock, imgRef, err := o.imagesLockIfIsBundle(throttleReq, image, processedImgs)
+			lNestedBundles, imgsLock, imgRef, err := o.imagesLockIfIsBundle(throttleReq, image, processedImgs, logger)
```

Comment:
what's `l` in `lNestedBundles`? should we just call it nestedBundles?

Reply:
done